lxd/storage/drivers/driver_common.go

```diff
@@ -515,3 +516,31 @@ func (d *common) roundVolumeBlockSizeBytes(sizeBytes int64) int64 {
 func (d *common) isBlockBacked(vol Volume) bool {
 	return vol.driver.Info().BlockBacking
 }
+
+// filesystemFreeze syncs and freezes a filesystem and returns an unfreeze function on success.
+func (d *common) filesystemFreeze(path string) (func() error, error) {
+	err := filesystem.SyncFS(path)
+	if err != nil {
+		return nil, fmt.Errorf("Failed syncing filesystem %q: %w", path, err)
+	}
+
+	_, err = shared.TryRunCommand("fsfreeze", "--freeze", path)
+	if err != nil {
+		return nil, fmt.Errorf("Failed freezing filesystem %q: %w", path, err)
+	}
+
+	d.logger.Info("Filesystem frozen", logger.Ctx{"path": path})
+
+	unfreezeFS := func() error {
+		_, err := shared.TryRunCommand("fsfreeze", "--unfreeze", path)
+		if err != nil {
+			return fmt.Errorf("Failed unfreezing filesystem %q: %w", path, err)
+		}
+
+		d.logger.Info("Filesystem unfrozen", logger.Ctx{"path": path})
+
+		return nil
+	}
+
+	return unfreezeFS, err
```

Comment:
`err` could just be `nil` here.

Reply:
Yeah I prefer it being hardcoded to nil, fixed.